source/codegen/metadata_validation.py

```diff
@@ -0,0 +1,287 @@
+from typing import Any, Dict, Iterable, List, Set
+from schema import Schema, And, Optional, Or, Use
+import common_helpers
+
+
+# Rules that can be suppressed
+class RULES:
+    # In general, arrays should have sizes passed in to the underlying C API.
+    ARRAY_PARAMETER_NEEDS_SIZE = "ARRAY_PARAMETER_NEEDS_SIZE"
+    # The gRPC layer can determine the size of input arrays, so we should not require
+    # callers to pass in the size. We should not add any new suppressions of this type.
+    INPUT_ARRAY_SHOULD_NOT_HAVE_PASSED_IN_SIZE = "INPUT_ARRAY_SHOULD_NOT_HAVE_PASSED_IN_SIZE"
+    # In general, enums shouldn't have duplicate values. This is helpful for catching typos
+    # although there are a few enums that have legitimate duplicates.
+    ENUMS_SHOULD_NOT_HAVE_DUPLICATE_VALUES = "ENUMS_SHOULD_NOT_HAVE_DUPLICATE_VALUES"
+
+
+def validate_metadata(metadata: dict):
+    try:
+        for function_name in metadata['functions']:
+            validate_function(function_name, metadata)
+        for attribute_group in common_helpers.get_attribute_groups(metadata):
+            for attribute_id in attribute_group.attributes:
+                validate_attribute(attribute_group.attributes[attribute_id], metadata)
+        function_enums = get_function_enums(metadata['functions'])
+        attribute_enums = get_attribute_enums(metadata)
+        used_enums = function_enums.union(attribute_enums)
+        for enum_name in metadata['enums']:
+            validate_enum(enum_name, used_enums, metadata)
+    except Exception as e:
+        raise Exception(
+            f"Failed to validate {metadata['config']['namespace_component']}") from e
+
+DOCUMENTATION_SCHEMA = Schema(
+    {
+        "description": str,
+        Optional("note"): str,
+        Optional("table_body"): list,
+        Optional("caution"): str,
+    }
+)
+
+SIZE_SCHEMA = Schema(
+    {
+        'mechanism': And(str, lambda s: not common_helpers.is_unsupported_size_mechanism_type(s)),
+        'value': Or(str, int),
+        Optional('value_twist'): str,
+        Optional('tags'): [str],
+    },
+)
+
+PARAM_SCHEMA = Schema(
+    {
+        'direction': And(str, lambda s: s in ('in', 'out')),
+        'name': str,
+        Optional('type'): str,
+        Optional('grpc_type'): str,
+        Optional('documentation'): DOCUMENTATION_SCHEMA,
+        Optional('enum'): str,
+        # This should be SIZE_SCHEMA but don't want to validate if codegen_method is no
```

Comment:
You have this a couple of places and it seems consequential to me. This is adding quite a few lines-of-code to the validate methods.

Can the conditional move up to the top-level?

I was first thinking that you would handle this with an Or syntax.

`Functions: [Or(FUNCTION_SCHEMA, NO_CODEGEN_SCHEMA)]`
where NO_CODEGEN_SCHEAM would constrain `codegen _method` to `no` and validate nothing else. And `FUNCTION_SCHEMA` would have the opposite constraint and be more strict.

But the way it's written seems more imperative than that. What about swapping:

```python
        FUNCTION_SCHEMA.validate(function)
        if function.get('codegen_method', 'public') != 'no':
```

with

```python
        if function.get('codegen_method', 'public') != 'no':
          FUNCTION_SCHEMA.validate(function)
```

Would that allow moving size and params to schema constraints instead of custom code?

It looks like part of the purpose of the custom validation code is to give better errors (though maybe they're pretty good by default?).

Reply:
I think something like using NO_CODEGEN_SCHEMA would work, but if we're not going to validate anything else anyway I think it makes sense to just skip it like the example you gave. I think I can make this work.